[2.9 ACRA] Using WebView from more than one process at once with the same data directory is not supported

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/6ba3eddd-911e-4b40-bef3-262118a3f189

```
java.lang.RuntimeException: Using WebView from more than one process at once with the same data directory is not supported. https://crbug.com/558377
at org.chromium.android_webview.AwBrowserProcess.a(PG:40)
at vo.c(PG:88)
at vo.b(PG:152)
at vo.a(PG:133)
at com.android.webview.chromium.WebViewChromiumFactoryProvider.a(PG:239)
at com.android.webview.chromium.WebViewChromium.init(PG:44)
at android.webkit.WebView.<init>(WebView.java:678)
at android.webkit.WebView.<init>(WebView.java:604)
at android.webkit.WebView.<init>(WebView.java:587)
at android.webkit.WebView.<init>(WebView.java:574)
at android.webkit.WebView.<init>(WebView.java:564)
at com.ichi2.anki.AbstractFlashcardViewer$MyWebView.<init>(AbstractFlashcardViewer.java:2)
at com.ichi2.anki.AbstractFlashcardViewer.createWebView(AbstractFlashcardViewer.java:1)
at com.ichi2.anki.AbstractFlashcardViewer.setInterface(AbstractFlashcardViewer.java:3)
at com.ichi2.anki.AbstractFlashcardViewer.displayCardQuestion(AbstractFlashcardViewer.java:3)
at com.ichi2.anki.Reviewer.displayCardQuestion(Reviewer.java:2)
at com.ichi2.anki.AbstractFlashcardViewer$NextCardHandler.displayNext(AbstractFlashcardViewer.java:11)
at com.ichi2.anki.AbstractFlashcardViewer$NextCardHandler.onProgressUpdate(AbstractFlashcardViewer.java:1)
at com.ichi2.async.DeckTask$TaskListener.onProgressUpdate(DeckTask.java:1)
at com.ichi2.async.DeckTask.onProgressUpdate(DeckTask.java:3)
at com.ichi2.async.DeckTask.onProgressUpdate(DeckTask.java:1)
at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:715)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:214)
at android.app.ActivityThread.main(ActivityThread.java:7050)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:965)
```

Specifically blames the `acra` process here

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ichi2.anki/com.ichi2.anki.CardTemplatePreviewer}: java.lang.RuntimeException: Using WebView from more than one process at once with the same data directory is not supported. https://crbug.com/558377 : Lock owner com.ichi2.anki:acra (pid 18647)
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3555)
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3707)
at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2220)
at android.os.Handler.dispatchMessage(Handler.java:107)
at android.os.Looper.loop(Looper.java:237)
at android.app.ActivityThread.main(ActivityThread.java:8016)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076)
Caused by: java.lang.RuntimeException: Using WebView from more than one process at once with the same data directory is not supported. https://crbug.com/558377 : Lock owner com.ichi2.anki:acra (pid 18647)
at org.chromium.android_webview.AwDataDirLock.b(chromium-TrichromeWebViewGoogle.apk-stable-410410683:26)
at ka.i(chromium-TrichromeWebViewGoogle.apk-stable-410410683:37)
at ka.b(chromium-TrichromeWebViewGoogle.apk-stable-410410683:15)
at ka.k(chromium-TrichromeWebViewGoogle.apk-stable-410410683:2)
at com.android.webview.chromium.WebViewChromiumFactoryProvider.h(chromium-TrichromeWebViewGoogle.apk-stable-410410683:2)
at com.android.webview.chromium.WebViewChromium.init(chromium-TrichromeWebViewGoogle.apk-stable-410410683:14)
at android.webkit.WebView.<init>(WebView.java:429)
at android.webkit.WebView.<init>(WebView.java:353)
at android.webkit.WebView.<init>(WebView.java:336)
at android.webkit.WebView.<init>(WebView.java:323)
at android.webkit.WebView.<init>(WebView.java:313)
at com.ichi2.anki.AbstractFlashcardViewer$MyWebView.<init>(AbstractFlashcardViewer.java:2)
at com.ichi2.anki.AbstractFlashcardViewer.createWebView(AbstractFlashcardViewer.java:1)
at com.ichi2.anki.AbstractFlashcardViewer.recreateWebView(AbstractFlashcardViewer.java:2)
at com.ichi2.anki.AbstractFlashcardViewer.setInterface(AbstractFlashcardViewer.java:2)
at com.ichi2.anki.AbstractFlashcardViewer.displayCardQuestion(AbstractFlashcardViewer.java:4)
at com.ichi2.anki.AbstractFlashcardViewer.displayCardQuestion(AbstractFlashcardViewer.java:1)
at com.ichi2.anki.CardTemplatePreviewer.displayCardQuestion(CardTemplatePreviewer.java:1)
at com.ichi2.anki.CardTemplatePreviewer.onCollectionLoaded(CardTemplatePreviewer.java:12)
at com.ichi2.anki.AnkiActivity.startLoadingCollection(AnkiActivity.java:4)
at com.ichi2.anki.CardTemplatePreviewer.onCreate(CardTemplatePreviewer.java:19)
at android.app.Activity.performCreate(Activity.java:7957)
at android.app.Activity.performCreate(Activity.java:7946)
at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1307)
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3530)
... 11 more
```

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/a6ecb7ec-4c4d-4af3-8baf-d6b1818aef38

--- a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
@@ -48,6 +48,7 @@ import com.ichi2.compat.CompatHelper;
 import com.ichi2.utils.LanguageUtil;
 import com.ichi2.anki.analytics.UsageAnalytics;
 import com.ichi2.utils.Permissions;
+import com.ichi2.utils.WebViewDebugging;
 
 import org.acra.ACRA;
 import org.acra.ReportField;
@@ -240,6 +241,16 @@ public class AnkiDroidApp extends MultiDexApplication {
         Timber.tag(TAG);
 
         Timber.d("Startup - Application Start");
+
+        // The ACRA process needs a WebView for optimal UsageAnalytics values but it can't have the same data directory.
+        // Analytics falls back to a sensible default if this is not set.
+        if (ACRA.isACRASenderServiceProcess() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+            try {
+                WebViewDebugging.setDataDirectorySuffix("acra");
+            } catch (Exception e) {
+                Timber.w(e, "Failed to set WebView data directory");
+            }
+        }
 
         // analytics after ACRA, they both install UncaughtExceptionHandlers but Analytics chains while ACRA does not
         UsageAnalytics.initialize(this);

--- a/AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java
@@ -33,7 +33,9 @@ import com.brsanthu.googleanalytics.request.EventHit;
 import com.ichi2.anki.AnkiDroidApp;
 import com.ichi2.anki.BuildConfig;
 import com.ichi2.anki.R;
+import com.ichi2.utils.WebViewDebugging;
 
+import org.acra.ACRA;
 import org.acra.util.Installation;
 
 import androidx.annotation.NonNull;
@@ -287,6 +289,20 @@ public class UsageAnalytics {
     }
 
 
+    protected static boolean canGetDefaultUserAgent() {
+        // #5502 - getDefaultUserAgent starts a WebView. We can't have two WebViews with the same data directory.
+        // But ACRA starts an :acra process which does not terminate when AnkiDroid is restarted. https://crbug.com/558377
+
+        // if we're not under the ACRA process then we're fine to initialize a WebView
+        if (!ACRA.isACRASenderServiceProcess()) {
+            return true;
+        }
+
+        // If we have a custom data directory, then the crash will not occur.
+        return WebViewDebugging.hasSetDataDirectory();
+    }
+
+
     /**
      * An Android-specific device config generator. Without this it's "Desktop" and unknown for all hardware.
      * It is interesting to us what devices people use though (for instance: is Amazon Kindle support worth it?
@@ -318,7 +334,7 @@ public class UsageAnalytics {
             // Instead they respond that they auto-parse User-Agent strings for analytics attribution
             // For maximum analytics built-in report compatibility we will send the official WebView User-Agent string
             try {
-                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
+                if (canGetDefaultUserAgent() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                     this.userAgent(WebSettings.getDefaultUserAgent(context));
                 } else {
                     this.userAgent(System.getProperty("http.agent"));

--- a/AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/WebViewDebugging.java
@@ -4,9 +4,13 @@ import android.content.SharedPreferences;
 import android.os.Build;
 import android.webkit.WebView;
 
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
 import androidx.annotation.UiThread;
 
 public class WebViewDebugging {
+
+    private static boolean sHasSetDataDirectory = false;
 
     @UiThread
     public static void initializeDebugging(SharedPreferences sharedPrefs) {
@@ -22,5 +26,17 @@ public class WebViewDebugging {
             boolean enableDebugging = sharedPrefs.getBoolean("html_javascript_debugging", false);
             WebView.setWebContentsDebuggingEnabled(enableDebugging);
         }
+    }
+
+    /** Throws IllegalStateException if a WebView has been initialized */
+    @RequiresApi(api = Build.VERSION_CODES.P)
+    public static void setDataDirectorySuffix(@NonNull String suffix) {
+        WebView.setDataDirectorySuffix(suffix);
+        sHasSetDataDirectory = true;
+    }
+
+    public static boolean hasSetDataDirectory() {
+        // Implicitly truth requires API >= P
+        return sHasSetDataDirectory;
     }
 }

